src/binary/decode.rs

```diff
@@ -0,0 +1,2179 @@
+use crate::ast::*;
+use id_arena::Id;
+use std::io::Read;
+
+/// A trait implemented by every Web IDL bindings thing that can be decoded from
+/// an input stream.
+pub(crate) trait Decode {
+    /// The output of decoding `Self` from the input stream.
+    ///
+    /// Since the custom section's arenas and AST nodes typically live inside
+    /// the `cx`, this is usually some sort of typed ID pointing into an arena.
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
+    where
+        Cx: Ids,
+        R: Read;
```

Comment:
FWIW the codegen and wonkiness of `Read` I've found is pretty unfortunate historically, could this perhaps just be `&mut &[u8]` or something like that?

Reply:
Yeah, I've seen this before too. I'll see what I can do.